[None][refactor] Simplify decoder state initialization

Funatiq opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Refactor**
  * Decoder request setup streamlined: external buffer manager removed from public calls and initialization split into smaller modular steps for clearer, more reliable per-call buffer handling.
  * Language bindings updated to reflect the simplified call signature.

* **New Features**
  * Added a public API to set beam width for individual decoder requests.

* **Chores**
  * Updated tests and internal call sites to match the new interfaces.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!--
Please write the PR title by following this template:

**[JIRA ticket/NVBugs ID/GitHub issue/None][type] Summary**

Valid ticket formats:
  - JIRA ticket: [TRTLLM-1234] or [FOOBAR-123] for other FOOBAR project
  - NVBugs ID: [https://nvbugs/1234567]
  - GitHub issue: [#1234]
  - No ticket: [None]

Valid types (lowercase): [fix], [feat], [doc], [infra], [chore], etc.

Examples:
  - [TRTLLM-1234][feat] Add new feature
  - [https://nvbugs/1234567][fix] Fix some bugs
  - [#1234][doc] Update documentation
  - [None][chore] Minor clean-up
-->

## Description

* Remove many parameters from decoder request and remove `newRequest` function.
* Instead introduce more fine grained functions to initialize decoder state directly.
* Now decoder request is only needed for speculative decoding initialization.

Follow up task:
* Simplify speculative decoding initialization and remove decoder request completely.

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

Funatiq committed:
refactor: Minor cleanup in CreateNewDecoderRequests

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: Move embedding bias initialization to a separate function in CreateNewDecoderRequests

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: Extract setupWords function in CreateNewDecoderRequests

- Moved the setupWords logic from a lambda expression into a dedicated function, improving code clarity and maintainability.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: Move newRequestSpeculativeDecoding out of newRequest

- Moved the setting of numDecodingEngineTokens to the newRequest function for better clarity.
- Reorganized speculative execution logic to improve readability and maintainability.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: Consolidate speculative decoding logic in createDecoderRequests

- Moved the speculative decoding logic for Medusa, Lookahead, and Eagle modes into a unified conditional block for improved clarity and maintainability.
- Ensured that the handling of decoder requests is streamlined and easier to follow.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: Enhance embedding bias initialization in CreateNewDecoderRequests

- Updated the initializeEmbeddingBias function to accept LlmRequest and logitsType, improving clarity and flexibility.
- Removed direct embedding bias handling from newRequest, streamlining the request creation process.
- Ensured that embedding bias is now consistently initialized using the updated function.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: Update stopWords and badWords handling in CreateNewDecoderRequests

- Removed badWordsList and stopWordsList from the request structure, streamlining the request parameters.
- Enhanced setupWords function to handle optional TensorPtr for requestWordsList, improving flexibility.
- Updated calls to setupWords in newRequest to utilize the new structure, ensuring proper memory management and GPU handling.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: Remove ids and endId from decoder_batch::Request

- Modified the Request constructor to remove the ids and endId parameters, streamlining the initialization process.
- Updated the newRequest function in CreateNewDecoderRequests to reflect changes in the Request structure, ensuring proper handling of input length and token generation.
- Removed unnecessary checks related to request IDs and endId, enhancing code clarity and maintainability.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: remove BufferManager from CreateNewDecoderRequests parameters

- Updated CreateNewDecoderRequests class and related methods to remove BufferManager as a parameter.
- CreateNewDecoderRequests already has runtimeStream as a parameter, so we can use it to create the BufferManager.
- This makes it easier to differentiate between the runtime stream and the decoder stream.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: Introduce initializeRequestIds and initializeBeamSearch functions in CreateNewDecoderRequests

- Added initializeRequestIds function to handle the initialization of request IDs and output IDs, improving code organization.
- Introduced initializeBeamSearch function to encapsulate beam search initialization logic, enhancing clarity and maintainability.
- Updated newRequest to utilize the new functions, streamlining the request handling process.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: Introduce initializeLogProbs function in CreateNewDecoderRequests

- Added initializeLogProbs function to encapsulate the initialization of cumulative and output log probabilities, improving code organization and clarity.
- Removed redundant log probability initialization logic from newRequest, streamlining the request handling process.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: Use decoder stream for request initialization

- Fix: Use decoder stream instead of runtime stream for request initialization.
- Updated functions in CreateNewDecoderRequests to utilize BufferManager instead of CudaStream for stream management, enhancing code clarity and consistency.
- Streamlined the initialization of request IDs, beam search, and embedding bias by passing BufferManager, improving memory management and GPU handling.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: Move beam width check to CreateNewDecoderRequests

- Move beam width check to CreateNewDecoderRequests from newRequest.
- Set beam width in CreateNewDecoderRequests instead of newRequest.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: Introduce initializeInputLengths function in CreateNewDecoderRequests

- Added initializeInputLengths function to encapsulate the logic for setting input lengths and validating constraints, improving code organization and clarity.
- Streamlined the newRequest function by replacing direct input length handling with a call to initializeInputLengths, enhancing maintainability.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: Initialize request lengths in CreateNewDecoderRequests

- Moved the initialization of input lengths from newRequest to the createDecoderRequests function.
- Removed the now unnecessary maxNewTokens parameter from the Request constructor, simplifying its initialization.
- Removed the now unnecessary maxSequenceLength parameter from the newRequest function.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: Rename newRequest to initializeOutputs

- All other functionality has been moved out of the newRequest function.
- Renamed newRequest to initializeOutputs to better reflect its purpose.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #13816](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13816/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #10387](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/10387/) completed with status: 'SUCCESS'

Funatiq committed:
refactor: Add setBeamWidth method to DecoderState

- Introduced setBeamWidth method in DecoderState to allow setting the beam width for specific requests in a batch.
- Updated CreateNewDecoderRequests to utilize the new setBeamWidth method, improving code clarity and maintainability.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #13857](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13857/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #13857](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13857/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #10425](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/10425/) completed with status: 'SUCCESS'
Pipeline passed with automatic retried tests. Check the [rerun report](https://urm.nvidia.com/artifactory/sw-tensorrt-generic/llm-artifacts//LLM/main/L0_MergeRequest_PR/10425/test-results/rerun_report.html) for details.

Funatiq commented:
/bot run
